core/silkworm/common/endian.hpp

```diff
@@ -185,31 +186,30 @@ ByteView to_big_compact(uint64_t value);
 //! \remarks A "compact" big endian form strips leftmost bytes valued to zero
 ByteView to_big_compact(const intx::uint256& value);
 
-//! \brief Parses unsigned integer from a compacted big endian byte form
-//! \param [in] data : byte view of compacted value. Length must be <= sizeof(UnsignedInteger)
-//! \param [in] allow_leading_zeros : when false, return std::nullopt if data starts with a 0 byte (i.e. not compact)
-//! \return The corresponding integer with native endianness; std::nullopt if data is invalid
-//! \remarks A "compact" big endian form strips leftmost bytes valued to zero
+//! \brief Parses unsigned integer from a compacted big endian byte form.
+//! \param [in] data : byte view of compacted value. Length must be <= sizeof(UnsignedInteger);
+//! otherwise kOverflow is returned.
+//! \param [out] out: the corresponding integer with native endianness.
+//! \return kOk or kOverflow or kLeadingZero.
+//! \remarks A "compact" big endian form strips leftmost bytes valued to zero;
+//! if the input is not compact kLeadingZero is returned.
 template <typename UnsignedInteger>
-static std::optional<UnsignedInteger> from_big_compact(ByteView data, bool allow_leading_zeros = false) {
+static DecodingResult from_big_compact(ByteView data, UnsignedInteger& out) {
     if (data.length() > sizeof(UnsignedInteger)) {
-        return std::nullopt;
+        return DecodingResult::kOverflow;
     }
 
-    UnsignedInteger x{0};
-
-    if (data.empty()) {
-        return x;
+    if (!data.empty() && data[0] == 0) {
+        return DecodingResult::kLeadingZero;
     }
 
-    if (data[0] == 0 && !allow_leading_zeros) {
-        return std::nullopt;
-    }
+    out = 0;
```

Comment:
Just a note ... This sets to zero all bytes of out.
Probably a `memset` to zero for all bytes NOT overwritten by subsequent `memcpy` depicts better the logic of the function but I understand it adds some overhead in computing.